### PR TITLE
Add openssl functions for use by TPM2 sealing calculations

### PR DESCRIPTION
**Pull request by ddstreet**
This is a partial PR from #28519 that includes only the openssl functions. Most of the new functions also include test cases.

### CODE REVIEW

**Comment by ddstreet** [comment]
looks like centos 8 doesn't support sha-1, i'll add check for unsupported digest algs

**Comment by ddstreet** [comment]
the jammy failure seems unrelated, if i'm reading it right i think the failure is:
`6717s qemu-system-x86_64: -chardev socket,id=chrtpm,path=/tmp/tmp.Ei0Sydnu7l/sock: Failed to connect to '/tmp/tmp.Ei0Sydnu7l/sock': No such file or directory`

**Comment by bluca** [comment]
It seems related to me:

`6714s modprobe: FATAL: Module tpm_ibmvtpm not found in directory /lib/modules/5.15.0-83-generic`

These tests are not failing in main currently

**Comment by mrc0mmand** [comment]
> It seems related to me:
> 
> `6714s modprobe: FATAL: Module tpm_ibmvtpm not found in directory /lib/modules/5.15.0-83-generic`
> 
> These tests are not failing in main currently

Could you please restart the amd64 job? I think the modprobe stuff is a red herring as I've seen the `Failed to connect to ...` error in CentOS CI as well a couple of times, and plan to address it soon-ish.

**Comment by ddstreet** [comment]
> `6714s modprobe: FATAL: Module tpm_ibmvtpm not found in directory /lib/modules/5.15.0-83-generic`

This happen in passing tests also, e.g.:

https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux/5918/artifact/systemd-centos-ci/artifacts_txiwv7qz/vagrant-logs.kzW/vagrant-arch-testsuite.ek5/TEST-70-TPM2_PASS.log

`modprobe: FATAL: Module tpm_ibmvtpm not found in directory /lib/modules/6.4.12-arch1-1
`

https://autopkgtest.ubuntu.com/results/autopkgtest-jammy-upstream-systemd-ci-systemd-ci/jammy/amd64/s/systemd-upstream/20230917_120603_d195a@/log.gz

`5252s modprobe: FATAL: Module tpm_ibmvtpm not found in directory /lib/modules/5.15.0-83-generic
`

Looking at where the module is loaded, it seems we load it in `test/TEST-70-TPM2/test.sh`, but looking at the kernel module it appears the hardware only exists on PPC:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/char/tpm/Kconfig#n171
so the driver won't get built on any other arch, which is why we always see the failure for non-PPC archs...I opened #29212 to restrict loading the module to ppc hw

**Comment by ddstreet** [comment]
rebased on main

File: src/shared/openssl-util.c [hunk 1]
```diff
@@ -681,28 +644,26 @@ int digest_and_sign(
...
-        r = openssl_hash(md_algorithm, s, len, hash, NULL);
+        _cleanup_free_ void *digest = NULL;
+        size_t digest_size;
+        r = openssl_digest(md_algorithm, s, len, &digest, &digest_size);
         if (r < 0)
                 return r;
 
-        enc = hexmem(hash, hash_size);
-        if (!enc)
+        *ret = hexmem(digest, digest_size);
```
Comment (bluca) [inline src/shared/openssl-util.c]: `ret` should not be overwritten in case of error, so keep the previous temp variable around and assign on success

File: src/test/test-openssl.c [hunk 2]
```diff
@@ -94,7 +94,7 @@ TEST(ecc_pkey_curve_x_y) {
         assert_se(memcmp_nn(y, y_len, y2, y2_size) == 0);
 }
 
-TEST(invalid) {
+TEST(invalid_pem) {
```
Comment (bluca) [inline src/test/test-openssl.c]: this change doesn't seem related to the rest of the commit? (`openssl: add openssl_digest_many()`)

File: src/shared/openssl-util.c [hunk 3]
```diff
@@ -236,6 +236,91 @@ int openssl_hmac_many(
         return 0;
 }
 
+/* Perform Key-Based HMAC KDF. The mode must be "COUNTER" or "FEEDBACK". The parameter naming is from the
```
Comment (bluca) [inline src/shared/openssl-util.c:407]: Add an assert for 'mode' to enforce this condition, so that we don't get it wrong in the future

**Review by bluca — APPROVED** [review]

File: src/test/test-openssl.c [hunk 4]
```diff
@@ -102,4 +102,56 @@ TEST(invalid) {
...
+TEST(digest_size) {
+        size_t size;
+        int r;
+
+        FOREACH_ARRAY(t, digest_size_table, ELEMENTSOF(digest_size_table)) {
+                _cleanup_free_ char *uppercase_alg = strdup(t->alg);
+
+                r = openssl_digest_size(t->alg, &size);
+                if (r == -EOPNOTSUPP)
+                        continue;
```
Comment (poettering) [inline src/test/test-openssl.c]: would probably be wise to generate some log output here, because it certainly is weird and unexpected if this fails on current openssl versions I guess.

File: src/shared/openssl-util.c [hunk 5]
```diff
@@ -236,6 +236,106 @@ int openssl_hmac_many(
...
+        }
+        if ((size_t) cipher_iv_length > iv_size)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Not enough IV bytes provided, require %d", cipher_iv_length);
+
+        if (!EVP_EncryptInit(ctx, cipher, key, iv))
+                return log_openssl_errors("Failed to initialize EVP_CIPHER_CTX.");
+
+        int cipher_block_size = EVP_CIPHER_CTX_block_size(ctx);
+
```
Comment (poettering) [inline src/shared/openssl-util.c:312]: maybe: `assert(cipher_block_size > 0)`

File: src/shared/tests.h [hunk 6]
```diff
@@ -43,7 +43,8 @@ bool can_memlock(void);
 #define DEFINE_HEX_PTR(name, hex)                                       \
         _cleanup_free_ void *name = NULL;                               \
         size_t name##_len = 0;                                          \
-        assert_se(unhexmem(hex, strlen(hex), &name, &name##_len) >= 0);
+        if (hex)                                                        \
+                assert_se(unhexmem(hex, strlen(hex), &name, &name##_len) >= 0);
```
Comment (poettering) [inline src/shared/tests.h]: instead: use strlen_ptr()

**Comment by poettering** [comment]
lgtm, just nitpicks

File: src/test/test-openssl.c (hunk 4 above)
Comment (ddstreet) [reply src/test/test-openssl.c]: Yeah, normally we should expect openssl to support all those algs, but it seems that older openssl (specifically what's in Centos Stream 8) doesn't support the alg naming that includes dashes; or at least, doesn't support some of them. For example:
https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos8/5819/artifact/systemd-centos-ci/artifacts_w0ttp9za/testsuite-logs-upstream.AQe/meson-logs/testlog.txt

> /* test_digest_size */
> Digest algorithm 'sha-1' not supported.
> Digest algorithm 'sha-224' not supported.
> Digest algorithm 'sha2-224' not supported.
> Digest algorithm 'sha-256' not supported.
> Digest algorithm 'sha2-256' not supported.
> Digest algorithm 'sha-384' not supported.
> Digest algorithm 'sha2-384' not supported.
> Digest algorithm 'sha-512' not supported.
> Digest algorithm 'sha2-512' not supported.
> Digest algorithm 'invalid.alg' not supported.
> 

It looks like upstream openssl added the names with dashes in this commit from 2019:
https://github.com/openssl/openssl/commit/df553b79419230d698d221919c7ceec68aa8c6c6

I think Centos Stream 8 is getting a bit old...how much longer do we need to stay compatible with it? Is it time to move up to Centos Stream 9? @mrc0mmand?

For this test, it actually might be better to remove the -EOPNOTSUPP check and actually fail if any of these algs aren't supported, and then just skip the test for Centos Stream 8?

File: src/test/test-openssl.c (hunk 2 above)
Comment (ddstreet) [reply src/test/test-openssl.c]: sorry! i think this slipped in while i was rebasing/editing; removed it.

File: src/shared/openssl-util.c (hunk 1 above)
Comment (ddstreet) [reply src/shared/openssl-util.c]: ack

File: src/shared/openssl-util.c (hunk 3 above)
Comment (ddstreet) [reply src/shared/openssl-util.c:407]: ack

File: src/shared/openssl-util.c (hunk 5 above)
Comment (ddstreet) [reply src/shared/openssl-util.c:312]: ack, added, thnx

File: src/shared/tests.h (hunk 6 above)
Comment (ddstreet) [reply src/shared/tests.h]: nice, updated, thanks

File: src/test/test-openssl.c (hunk 4 above)
Comment (ddstreet) [reply src/test/test-openssl.c]: ah, so it looks like this 2019 commit only made it into openssl 3, so i think the "sha-*" alg names can be ifdef'ed and we can then assume all the algs are supported
